libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp

```diff
@@ -778,6 +786,44 @@ uint32_t AP_Frsky_Telem::calc_attiandrng(void)
     return attiandrng;
 }
 
+/*
+ * prepare waypoint data
+ * for FrSky SPort Passthrough (OpenTX) protocol (X-receivers)
+*/
+uint32_t AP_Frsky_Telem::calc_wp(void)
+{
+    uint32_t waypoint;
+    // wp number, max is 1023 encoded as 10bits
+    waypoint = MIN(_nav_info.wp_number,WAYPOINT_NUMBER_LIMIT);
+    // max wp distance is 1023*100m = 102.3Km encoded as 12bits: 10 + 2 for 10^power
+    waypoint |= prep_number(MIN(_nav_info.wp_distance,WAYPOINT_DISTANCE_LIMIT),3,2) << WAYPOINT_DISTANCE_OFFSET;
+    // xtrack error, max is +/- 150m encoded as 6bits: 4 + 1 for 10^power + 1 for sign
+    waypoint |= prep_number(MIN(_nav_info.wp_xtrack_error,WAYPOINT_XTRACK_LIMIT),1,1) << WAYPOINT_XTRACK_OFFSET;
+    const bool gotGPS = (AP::gps().status() >= AP_GPS::GPS_OK_FIX_2D);
+    //
+    if (gotGPS) {
+        const int32_t cog = AP::gps().ground_course_cd(); //cog in centidegrees
```

Comment:
Are you assuming you're flying forward here?

Reply:
ok this is another good question.
As I said I took inspiration from OSD code 
https://github.com/ArduPilot/ardupilot/blob/c0af0cfe70a8a102b1d7dfa436a46d7c32039ce3/libraries/AP_OSD/AP_OSD_Screen.cpp#L849
but I noticed it was based on yaw which in my opinion did not make much sense for vehicles able to move while rotating. I thought that cog was a better choice and this is the result.
I did not make any assumption but since this is all new to me I might have missed the obvious!

Reply:
Again, I would look at the NAV_CONTROLLER_OUTPUT MAVLink message for inspiration (careful, as it is used differently depending on the vehicle, from what I understand of the ArduPilot code):
https://github.com/ArduPilot/mavlink/blob/master/message_definitions/v1.0/common.xml#L3245